### PR TITLE
Fix #1984

### LINKED ISSUE #1984: Config `newPlayersSpawnInTF` does not work properly under certain conditions
### NeoForge Version

20.4.140

### Twilight Forest Version

4.4.2156

### Client Log

_No response_

### Crash Report (if applicable)

_No response_

### Steps to Reproduce

In the config file, have `newPlayersSpawnInTF` set to true, but set `portalForNewPlayer` to false.

### What You Expected

Spawn in TF with your feet firmly planted on the ground.

### What Happened Instead

Spawn in TF with your feet in the process of being firmly planted in the ground through gravitational forces.

### Additional Details

I've been slapping some breakpoints around the place as well as teasing the configs. On accident, I tested the config with the portal being made alongside the player spawning in. This is where I find out that the logic works perfectly fine if `portalForNewPlayer` is true, but if it's set to false, the player will end up being spawned at least 64 blocks above surface. Simple observation: it's the Y coordinate for spawning in the overworld.

I put breakpoints around `TFTeleporter.getPortalInfo` and found this while `destPos` is initially incorrect, this coordinate does get corrected in `TFTeleporter.makePortal`, but only in the variables. While `spot` does get corrected in `TFTeleporter.findPortalCoords`, the problem is that the portal coords never get cached in `TFTeleporter.cacheNewPortalCoords`. The simple reason is that the cache is trying to look for portal blocks, but since we spawned in without the portal, the coords are never cached, leading to a cascade back up to the second call for `TFTeleporter.placeInExistingPortal` where the cache is empty, meaning `blockpos` returns null, meaning `ITeleporter.super.getPortalInfo` is returned, rather than the pos. As such, the coordinates to spawn in never get set properly and relies on the overworld's spawn point.

### Please Read and Confirm The Following

- [X] I have confirmed this bug can be replicated without the use of Optifine.
- [X] I have confirmed the details provided in this report are concise as possible and does not contained vague information (ie. Versions are properly recorded, answers to questions are clear).
- [X] I have confirmed this issue is unique and has not been reported already.

### PR DESCRIPTION
**Pull request by Andromander**
Doing a few runs with the specified config setup, this appears to mostly work. There was one instance I spawned in the Dark Forest, but I'm sure that's just because I was in Creative and did not have any Progression blocking me.

I haven't committed this because it's a little...ugly. Works, just doesn't look good. For one, NoReturnTeleporter now overrides `getPortalInfo` from TFTeleporter. Not terrible, but does copy a lot, so I feel it could just be in TFTeleporter. Second, a lot of things had to be made protected, since TFTeleporter had all the logic. I also made a modified version of `makePortal` that returns a PortalInfo so that the coordinates used there are directly used to set the position of the player.

TL;DR: I fixed, but I think it should be all in TFTeleporter.

### CODE REVIEW

**Comment by Tamaized** [comment]
Can we abstract this instead of duplicating logic between the two classes?

**Comment by Andromander** [comment]
Basically the gist of me creating this ticket because I feel like it should be possible, just that I couldn't really see it.

One reason I did duplicate the logic was because the only use for it was banishing new players into TF, rather than it being used to teleport players frequently. It'd need an additional parameter to tell whether it should teleport and cache a portal, or just teleport. Probably possible to do, anyway.

**Comment by Andromander** [comment]
I'm going to pull this as is for now since the aim was to fix an issue, but further discussion shows the teleporter/s need some reworking. Since I believe that's out of scope for this and was only spurred by my annoyance with my own code, this will do for now.
